WIP: k-shortest path and new rounding

NinaWie opened the pull request.

NinaWie commented on ebike_city_tools/optimize/round_optimized_sort_selection.py:
```diff
@@ -0,0 +1,224 @@
+def rounding_error(fractional_value):
+    """Given a fractional value, returns the distance to the closest integer."""
+    next_int = round(fractional_value)
+    if fractional_value >= 0:
```
I don't understand this line - the fractional_value should always be greater 0, shouldn't it? So this condition is usually true, so the `rounding_error` method usually yields 0?

CNoebel replied:
Ah yeah it should be a >= 1. That was one change I tried out where we allocate at most one lane for cars

NinaWie commented on ebike_city_tools/optimize/round_optimized_sort_selection.py:
```diff
@@ -0,0 +1,224 @@
+def round_row(row):
+    """Given a row, this function returns the row obtained by rounding the car and bike capacities."""
+    # print(row)
+    new_value = min(1, round(row["u_c(e)"]))
```
So u_c always stays at least 1 - isn't that too restrictive?

CNoebel replied:
It is at most 1. The idea is, that in our model having two parallel arcs does not improve the car metric.

NinaWie replied:
Sorry, I always get confused with min - this makes sense.

NinaWie commented on ebike_city_tools/optimize/round_optimized_sort_selection.py:
```diff
@@ -0,0 +1,224 @@
+def sort_by_rounding_error(capacities) :
+    """Sorts the capacities by increasing rounding error of the car capacity."""
+    cap_with_rounding_error = capacities
+    cap_with_rounding_error["rounding_error"] = capacities.apply(rounding_error_of_row, axis=1)
```
Could probably be accelarated by doing something like 
```
capacities["rounded_car"] = capacities["u_c(e)"].round()
cap_with_rounding_error["rounding_error"] = (capacities["u_c(e)"] - capacities["rounded_car"]).abs()
```
(.apply is usually slower than column-wise operations)

CNoebel replied:
Sounds great! I do not know, how to write good pandas code ^^'

NinaWie replied:
I can make such changes after we merge the code

NinaWie commented on ebike_city_tools/optimize/round_optimized_sort_selection.py:
```diff
@@ -0,0 +1,224 @@
+            - "highest_bike_value": Rounds up the highest bike value to a bike lane, if possible by connectivity.
+            - "lowest_rounding_error": Rounds the car value closest to an integer, if possible by connectivity.
+        """
+        self.od = extend_od_circular(od, list(G_lane.nodes()))
```
This should probably be done in the `compare_rounding.py` script, to make sure that the same OD matrix (with extension) is used for all algorithms. Well spotted though that this was missing! My bad, I usually included it in the synthetic-data-generation part in each script, but seems like I forgot it here

CNoebel replied:
Oh I did not add it, I copy pasted this part from the Pareto Round Optimzer

NinaWie replied:
Ah, I got confused here - I added it to the Pareto Optimizer because the OD matrix should not be extended for the betweenness centrality algorithms (only for ours). So we should keep it as is.

NinaWie commented on ebike_city_tools/optimize/round_optimized_sort_selection.py:
```diff
@@ -0,0 +1,224 @@
+        return output_to_dataframe(ip, self.G_street, fixed_edges=fixed_capacities)
+
+
+    def pareto(self) -> pd.DataFrame:
```
To make the code prettier, we could try to make one ParetoOptimize class and inherit from it :)

NinaWie replied:
I just realized that you already wrote it in a way that this code would replace my code, which is even better :)

CNoebel replied:
I tried, and it gives the same Pareto front as the other code. Using inheritance could be better though, as there are some fundamental differences that lead to weird if clauses.

NinaWie replied:
I think it's not too bad since it's mainly one if clause. Inheritance would only remove the duplication of `init` and `optimize` functions, but not the long pareto function, so I actually prefer this version. After merging, I will delete the other pareto_optimize script. great that you tested that it's the same

NinaWie commented:
@CNoebel , sorry for all the comments, most of them are just random thoughts, but I might have found a bug in the `rounding_error` method - If you further test your method, maybe have a look at that :)

NinaWie commented:
Short update, I made the changes from my comments (simplified the rounding and changed >=0 to >=1) and adapted the compare_rounding script to the newest version.

NinaWie commented:
Further update on the k shortest path thing: I tested your implementation in a separate script (now added as `scripts/compare_kshortestpaths.py`), similar to my tests for the OD matrix size. The implementation works, but it reduces the number of considered edges by a lot, leading to significantly worse objective values. The problem is that after deleting the shortest paths, there is oftentimes just 1 or two more possible paths that can even be found. This makes sense since most nodes only have ~3 neighboring nodes, so after removing the two shortest paths, there is no neighbor left.

I now implemented another version of sub-selecting valid edges: given the shortest path P between s and t, take all nodes on P, and select the k nearest nodes, where "nearest" refers to their distacne in *geographic* space. This is also not perfect, but allows for some more flexibility in how much edges are selected. I computed the runtimes and increase in objective value for both versions, see attached picture. It seems that decreases of around 50% in runtime are possible for the tested graphs without making the objective value increase a lot. I'm now testing it also for bigger graphs.

[kshortest_scatter_christian.pdf](https://github.com/mie-lab/bike_lane_optimization/files/13872955/kshortest_scatter_christian.pdf)
[kshortest_scatter_nina.pdf](https://github.com/mie-lab/bike_lane_optimization/files/13872956/kshortest_scatter_nina.pdf)

NinaWie commented:
I would merge this branch today, except you see any issues with that @CNoebel ?
